File: 0832. Flipping an Image/0832.java
// time complexity: O(n^2)
// space complexity: O(1)
class Solution {
    public int[][] flipAndInvertImage(int[][] image) {
        // n x n image (square)
        int n = image.length;

        // first, we should flip horizontally
        for(int i=0; i<n; i++){
            int temp = n-1;
            for(int j=0; j<n/2; j++){
                //swap(image[i][j], image[i][temp]);
                image[i][j] = image[i][j] ^ image[i][temp];
                image[i][temp] = image[i][j] ^ image[i][temp];
                image[i][j] =image[i][j] ^ image[i][temp];
                temp--;
            }
        }

        // then, invert it.
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                image[i][j] ^= 1;
            }
        }

        return image;
    }
}

File: 0832. Flipping an Image/0832.py
# time complexity: O(n^2), reversing + iterating each bit
# space complexity: O(n^2), for the new matrix 
class Solution(object):
    def flipAndInvertImage(self, image):
        """
        :type image: List[List[int]]
        :rtype: List[List[int]]
        """
        # for each row: reverse it then invert each bit by 1−bit
        return [[1 - bit for bit in row[::-1]] for row in image]
        # row[::-1]] -> reads the row list from start to finish